src/python/pants/engine/target_test.py

```diff
@@ -0,0 +1,186 @@
+                    product_type=Snapshot,
+                    subject_type=PathGlobs,
+                    mock=lambda _: Snapshot(
+                        directory_digest=EMPTY_DIRECTORY_DIGEST,
+                        files=hydrated_source_files,
+                        dirs=(),
+                    ),
+                )
+            ],
+        )
+        return result
+
+    # Normal field
+    expected_files = ("monad.hs", "abstract_art.hs", "abstract_algebra.hs")
+    assert (
+        hydrate_field(
+            raw_source_files=["monad.hs", "abstract_*.hs"], hydrated_source_files=expected_files
+        ).snapshot.files
+        == expected_files
+    )
+
+    # Test pre-hydration validation
+    with pytest.raises(ValueError) as exc:
+        hydrate_field(raw_source_files=[0, 1, 2], hydrated_source_files=())  # type: ignore[call-arg]
+    assert "Not all elements of the iterable have type" in str(exc)
+
+    # Test post-hydration validation
+    with pytest.raises(ValueError) as exc:
+        hydrate_field(raw_source_files=["*.js"], hydrated_source_files=("not_haskell.js",))
+    assert "Received non-Haskell sources" in str(exc)
+
+
+def test_has_fields() -> None:
+    class UnrelatedField(PrimitiveField):
+        alias: ClassVar = "unrelated"
+        raw_value: Optional[bool]
+
+        @memoized_property
+        def value(self) -> bool:
+            if self.raw_value is None:
+                return False
+            return self.raw_value
+
+    tgt = HaskellTarget({})
+    assert tgt.has_fields([]) is True
+    assert tgt.has_fields([HaskellGhcExtensions]) is True
+    assert tgt.has_fields([UnrelatedField]) is False
+    assert tgt.has_fields([HaskellGhcExtensions, UnrelatedField]) is False
+
+
+def test_field_hydration_is_lazy() -> None:
+    # No error upon creating the Target because validation does not happen until a call site
+    # hydrates the specific field.
+    tgt = HaskellTarget(
+        {HaskellGhcExtensions.alias: ["GhcExistentialQuantification", "DoesNotStartWithGhc"]}
+    )
+    # When hydrating, we expect a failure.
+    with pytest.raises(ValueError) as exc:
+        tgt.get(HaskellGhcExtensions).value
+    assert "must be prefixed by `Ghc`" in str(exc)
+
+
+def test_add_custom_fields() -> None:
+    class CustomField(PrimitiveField):
+        alias: ClassVar = "custom_field"
+        raw_value: Optional[bool]
+
+        @memoized_property
+        def value(self) -> bool:
+            if self.raw_value is None:
+                return False
+            return self.raw_value
+
+    union_membership = UnionMembership(OrderedDict({HaskellTarget: OrderedSet([CustomField])}))
+    tgt_values = {CustomField.alias: True}
+    tgt = HaskellTarget(tgt_values, union_membership=union_membership)
+    assert tgt.field_types == (HaskellGhcExtensions, HaskellSources, CustomField)
+    assert tgt.core_fields == (HaskellGhcExtensions, HaskellSources)
+    assert tgt.plugin_fields == (CustomField,)
```

Comment:
Nice.

Re: naming: `plugin` is slightly ill-defined, and a good plugin model mostly erases the boundaries between plugins and non-plugins, making everything feel first class. That might affect naming here, but you might also want to hide the `plugin_fields` accessor and just expose "default_fields" and "fields".

Reply:
See https://github.com/pantsbuild/pants/pull/9284/files#r391961580 on why I think there's value in still having `plugin_fields`. tl;dr, we expect to almost always use `tgt.field_types`, and the Target code will always use that. But, it helps with debugging to know what was set up by default vs. through your plugin.